game-app/game-core/src/main/java/games/strategy/triplea/ui/CasualtySelection.java

```diff
@@ -0,0 +1,189 @@
+import java.awt.Component;
+import java.awt.Dimension;
+import java.awt.Font;
+import java.awt.Toolkit;
+import java.math.BigDecimal;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.swing.BorderFactory;
+import javax.swing.JDialog;
+import javax.swing.JOptionPane;
+import javax.swing.JScrollPane;
+
+/**
+ * drives the casualty selection process:
+ *
+ * <p>1. construct a {@code CasualtySelection} object e.g. {@code final var casualtySelection = new
+ * CasualtySelection([...]);}
+ *
+ * <p>2. call {@code showModelDialog} e.g.
+ * {@code final CasualtyDetails selectedCasualties =
+ *    casualtySelection.showModalDialog().orElse(null);}
+ */
+
+public class CasualtySelection {
+  @VisibleForTesting
+  public UnitChooser chooser;
+
+  private final int hitsToTake;
+  private final GamePlayer player;
+
+  private final JDialog dialog;
+  private final boolean isEditMode;
+
+  private final JOptionPane optionPane;
+
+  public CasualtySelection(final Collection<Unit> selectFrom,
+                           final Map<Unit, Collection<Unit>> dependents,
+                           final int hitsToTake,
+                           final String title,
+                           final GamePlayer player,
+                           final CasualtyList defaultCasualties,
+                           final boolean allowMultipleHitsPerUnit,
+                           final UiContext uiContext,
+                           final Component dialogParent,
+                           final boolean isEditMode) {
+    this.hitsToTake = hitsToTake;
+    this.player = player;
+
+    this.isEditMode = isEditMode;
+
+    final boolean movementForAirUnitsOnly =
+        playerMayChooseToDistributeHitsToUnitsWithDifferentMovement(selectFrom);
+
+    chooser =
+        new UnitChooser(
+            selectFrom,
+            defaultCasualties,
+            dependents,
+            Properties.getPartialAmphibiousRetreat(
+                player.getData().getProperties()),
+            movementForAirUnitsOnly,
+            allowMultipleHitsPerUnit,
+            uiContext);
+    chooser.setTitle(title);
+
+    if (isEditMode) {
+      chooser.disableMax();
+    } else {
+      chooser.setMax(hitsToTake);
+    }
+
+          final JScrollPane chooserScrollPane = new JScrollPane(chooser);
+    chooserScrollPane.setBorder(BorderFactory.createEmptyBorder(1, 1, 1, 1));
+    optionPane = new JOptionPane(
+        chooserScrollPane,
+        JOptionPane.PLAIN_MESSAGE,
+        JOptionPane.OK_CANCEL_OPTION,
+        null,
+        null, // options,
+        JOptionPane.OK_OPTION);
+
+    dialog = optionPane.createDialog(dialogParent, player.getName() + " select casualties");
+  }
+
+  /**
+   * @return the CasualtyDetails that describe which casualties the player selected
+   * or null iff something went weirdly wrong
```

Comment:
This comment indicates a `null` could be returned, it probably should say `empty Optional` instead. A null value would be possible to return, it's a major red flag though if `null` is ever returned instead of an empty Optional.

Reply:
Comment will be corrected in the next commit